html/mycakeapp/src/Controller/MypageController.php

```diff
@@ -42,4 +45,135 @@ public function delete()
+        $memberId = $this->Auth->user('id');
+        $cardsInfoOwn = $this->Creditcards->find('CardsInfoOwn', ['memberId' => $memberId]);
+        $securityKey = Configure::read('key');
+        $securitySalt = Configure::read('salt');
+        foreach ((array)$cardsInfoOwn as $key => $value) {
+            $lastFourDigits = substr($value['card_number'], -4, 4);
+            $encryptedCardId = Security::encrypt($value['id'], $securityKey, $securitySalt);
+            $cardsInfoOwn[$key]['card_number'] = $lastFourDigits;
+            $cardsInfoOwn[$key]['id'] = $encryptedCardId;
+        }
+        $title = '決済情報';
+        $this->set(compact('title', 'cardsInfoOwn'));
+    }
+    public function addpayment()
+    {
+        $this->viewBuilder()->setLayout('frame-title');
+        $entity = $this->Creditcards->newEntity();
+        $memberId = $this->Auth->user('id');
+        $cardsInfoOwn = $this->Creditcards->find('CardsInfoOwn', ['memberId' => $memberId]);
+        if (!empty($this->request->query['id'])) { //updateの$entity
+            // 暗号化したクレジットカードIDを戻す
+            $securityKey = Configure::read('key');
+            $securitySalt = Configure::read('salt');
+            $encryptedCardId = $this->request->query['id'];
+            $cardId = Security::decrypt($encryptedCardId, $securityKey, $securitySalt);
+            if ($cardId === false) { //不正なIDの時
+                return $this->redirect(['controller' => 'error']);
+            }
+            foreach ((array)$cardsInfoOwn as $cardInfoOwn) {
+                if ($cardInfoOwn['id'] === (int)$cardId) {
+                    $cardHolderIsMember = true;
+                }
+            }
+            if (empty($cardHolderIsMember)) { //カードがユーザーのものじゃない時
+                return $this->redirect(['controller' => 'error']);
+            }
+            $entity = $this->Creditcards->get($cardId);
+            $previousCardNumber = $entity['card_number'];
+        } else { //insertのカード数判定
+            $numberOfCardsOwned = count($cardsInfoOwn);
+            if ($numberOfCardsOwned === 2) {
+                return $this->redirect(['controller' => 'error']);
+            }
+        }
+        if (!empty($this->request->is('Put'))) { //update
+            $entity = $this->Creditcards->patchEntity($entity, $this->request->getData());
+
+            $isCardNumberExist = in_array($entity['card_number'], $this->Creditcards->find('AllCardNumbers'), true);
+            if ($isCardNumberExist === true && $entity['card_number'] !== $previousCardNumber) {
+                $cardNumberIsNotUnique = "このクレジットカードは利用できません";
+                $this->set(compact('cardNumberIsNotUnique'));
+            }
+            $entity["updated_at"] = date("Y/m/d H:i:s");
+            if (empty($cardNumberIsNotUnique) && $this->Creditcards->save($entity)) {
+                $_SESSION['addedpayment'] = 1;
+                return $this->redirect(['action' => 'addedpayment']);
+            }
+        }
```

Comment:
今のままだと、同じクレジットカード番号を登録したときに、エラー文は表示されるがDBに登録されてしまう状態になっています。
```suggestion
            if ($isCardNumberExist === true && $entity['card_number'] !== $previousCardNumber) {
                $cardNumberIsNotUnique = "このクレジットカードは利用できません";
                $this->set(compact('cardNumberIsNotUnique'));
            } else {
                $entity["updated_at"] = date("Y/m/d H:i:s");
                if (empty($cardNumberIsNotUnique) && $this->Creditcards->save($entity)) {
                    $_SESSION['addedpayment'] = 1;
                    return $this->redirect(['action' => 'addedpayment']);
                }
            }
```
こうすることで、エラー文があるときにsaveを通ることがなくなるので対応お願いいたします！